pkg/scheduler/scheduler.go

```diff
@@ -88,197 +47,231 @@ func NewScheduler(
-			return
+func (s *Scheduler) mainLoop(ctx context.Context, signals <-chan os.Signal) {
+	s.logger.Info("scheduler-started")
+	defer s.logger.Info("scheduler-stopped")
+
+	timer := time.NewTimer(0)
+	stoppedWorker := make(chan workerID, 1)
+
+	defer func() {
+		for {
+			if len(s.workers) == 0 {
+				return
+			}
+			timer := time.NewTimer(30 * time.Second)
+			select {
+			case id := <-stoppedWorker:
+				s.workersRunning.Done()
+				delete(s.workers, id)
+			case <-timer.C:
+				s.logger.Info("timeout_waiting_for_workers", lager.Data{"instances": s.ListIntanceGUIDs()})
+			}
 		}
+	}()
+
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(time.Duration(s.instanceRefreshInterval) * time.Second)
+
+			instanceInfos, err := s.brokerinfo.ListInstances()
+			if err != nil {
+				s.logger.Error("unable to retreive instance guids", err)
+				continue
+			}
+
+			s.logger.Debug("refresh_instances", lager.Data{"instances": instanceInfos})
 
-		w.logger.Debug("refresh_instances", lager.Data{"instances": instanceInfos})
-		for _, instanceInfo := range instanceInfos {
-			for driverName, driver := range w.metricsCollectorDrivers {
-				if utils.SliceContainsString(driver.SupportedTypes(), instanceInfo.Type) {
-					id := workerID{Driver: driverName, InstanceGUID: instanceInfo.GUID}
-					if _, ok := w.workers.get(id); !ok {
-						w.startWorker(id, instanceInfo)
+			desiredWorkerIDs := map[workerID]brokerinfo.InstanceInfo{}
+			for _, instanceInfo := range instanceInfos {
+				for driverName, driver := range s.metricsCollectorDrivers {
+					if utils.SliceContainsString(driver.SupportedTypes(), instanceInfo.Type) {
+						id := workerID{Driver: driverName, InstanceGUID: instanceInfo.GUID}
+						desiredWorkerIDs[id] = instanceInfo
 					}
 				}
 			}
-		}
 
-		// Stop any instance not returned by the brokerInfo which has a worker
-		for _, workerID := range w.workers.keys() {
-			stillExists := false
-			for _, instanceInfo := range instanceInfos {
-				if instanceInfo.GUID == workerID.InstanceGUID {
-					stillExists = true
-					break
+			for id, instanceInfo := range desiredWorkerIDs {
+				if _, ok := s.workers[id]; !ok {
+					workerContext, workerCancel := context.WithCancel(ctx)
+					worker := &collectorWorker{
+						id:             id,
+						instanceInfo:   instanceInfo,
+						driver:         s.metricsCollectorDrivers[id.Driver],
+						metricsEmitter: s.metricsEmitter,
+						retryInterval:  s.collectorRetryInterval,
+						maxRetries:     s.collectorMaxRetries,
+						cancel:         workerCancel,
+						logger:         s.logger,
+					}
+					s.workers[id] = worker
+					s.workersRunning.Add(1)
+					go worker.run(workerContext, stoppedWorker)
 				}
 			}
-			if !stillExists {
-				w.stopWorker(workerID)
+
+			for id, worker := range s.workers {
+				if _, ok := desiredWorkerIDs[id]; !ok {
+					worker.cancel()
+				}
 			}
+		case id := <-stoppedWorker:
+			s.workersRunning.Done()
```

Comment:
we reuse these two lines in two places, I would say it warrants a separate method (deleteWorker)